src-utils/HydraAuctionUtils/Monads.hs

```diff
@@ -0,0 +1,110 @@
+module HydraAuctionUtils.Monads (
+  UtxoQuery (..),
+  MonadTrace (..),
+  MonadSubmitTx (..),
+  MonadQueryUtxo (..),
+  MonadNetworkId (..),
+  MonadCardanoClient,
+  logMsg,
+  submitAndAwaitTx,
+  fromPlutusAddressInMonad,
+  addressAndKeysForActor,
+) where
+
+-- Prelude imports
+import Prelude
+
+-- Cardano imports
+import Cardano.Api.UTxO qualified as UTxO
+import Hydra.Cardano.Api (
+  Address,
+  AddressInEra,
+  NetworkId,
+  PaymentKey,
+  ShelleyAddr,
+  SigningKey,
+  Tx,
+  TxIn,
+  VerificationKey,
+  fromPlutusAddress,
+  getTxId,
+  txBody,
+ )
+
+-- Plutus imports
+import Plutus.V1.Ledger.Address qualified as PlutusAddress
+
+-- HydraAuction imports
+import CardanoClient (buildAddress)
+import Control.Monad.IO.Class (MonadIO (liftIO))
+import HydraAuctionUtils.Extras.CardanoApi (networkIdToNetwork)
+import HydraAuctionUtils.Fixture (Actor, keysFor)
+
+-- MonadQueryUtxo
+
+data UtxoQuery
+  = ByAddress (Address ShelleyAddr)
+  | ByTxIns [TxIn]
+
+class Monad m => MonadQueryUtxo m where
+  queryUtxo :: UtxoQuery -> m UTxO.UTxO
+
+-- MonadNetworkId
+
+class Monad m => MonadNetworkId m where
+  askNetworkId :: m NetworkId
+
+fromPlutusAddressInMonad ::
+  MonadNetworkId m => PlutusAddress.Address -> m AddressInEra
+fromPlutusAddressInMonad address = do
+  networkId <- askNetworkId
+  let network = networkIdToNetwork networkId
+  return $
+    fromPlutusAddress network address
+
+addressAndKeysForActor ::
+  (MonadNetworkId m, MonadIO m) =>
+  Actor ->
+  m
+    ( Address ShelleyAddr
+    , VerificationKey PaymentKey
+    , SigningKey PaymentKey
+    )
+addressAndKeysForActor actor = do
+  networkId' <- askNetworkId
+  (actorVk, actorSk) <- liftIO $ keysFor actor
+  let actorAddress = buildAddress actorVk networkId'
+  pure (actorAddress, actorVk, actorSk)
+
+-- MonadTrace
+
+class Monad m => MonadTrace m where
+  type TracerMessage m = mt | mt -> m
```

Comment:
this means that if you know the trace message you always know the Monad? does that make sense? I don't see that e.g. for `String`s there can only be one valid `Monad` to trace in

Reply:
by chance, I wrote a similar thing for my current issue: 
```haskell
class MonadTracer t m | m -> t where
  trace :: t -> m ()

-- | a transformer that provides a tracer
newtype TracerT t m a = TracerT {runTracerT :: ReaderT (Tracer m t) m a}
  deriving newtype
    ( Functor
    , Applicative
    , Monad
    , MonadReader (Tracer m t)
    , MonadIO
    )

-- these could be derived in a normal deriving clause but I feel better specifying
-- the context myself
deriving newtype instance MonadWriter w m => MonadWriter w (TracerT t m)
deriving newtype instance MonadState s m => MonadState s (TracerT t m)

mapTracerT :: (t1 -> t2) -> TracerT t1 m a -> TracerT t2 m a
mapTracerT f = TracerT . withReaderT (contramap f) . runTracerT

runWithTracer :: (forall b. n b -> m b) -> Tracer n t -> TracerT t m a -> m a
runWithTracer natTrans tracer = flip runReaderT (natTracer natTrans tracer) . runTracerT

runWithTracer' :: Tracer m t -> TracerT t m a -> m a
runWithTracer' = runWithTracer id

askTracer :: forall t m. Monad m => TracerT t m (Tracer m t)
askTracer = ask @(Tracer m t) 
```

Reply:
(yes `TracerT` is not really that relevant, except for their implementation of `MonadTrace` is I guess similar to the one of your ReaderT IO newtype)

Reply:
> does that make sense? I don't see that e.g. for Strings there can only be one valid Monad to trace in

In all places we use tracer they always define custom log datatypes, like HydraAuctionLog or EndToEndLog. Do not know why so, but this is current state of affairs.

> TracerT

I think we wont need a separate transformer, cuz all Runners have custom context which may have tracer.

Reply:
> I think we wont need a separate transformer, cuz all Runners have custom context which may have tracer.

my thing uses this so that's that

Reply:
> In all places we use tracer they always define custom log datatypes, like HydraAuctionLog or EndToEndLog. Do not know why so, but this is current state of affairs.

that's not the point though, you really will not be able to say 
`instance MonadTrace SomeMonad` and `instance MonadTrace SomeOtherMonad` where the log type is not fixed, you will always have to choose one specific log type, so that completely defeats the purpose because it can never be polymorphic.

Reply:
if you have a log type, you will have a fixed Monad, in that case you can just monomorphise...